src/webapp/context-providers/CurrentModuleContextProvider.tsx

```diff
@@ -0,0 +1,79 @@
+import { useEffect, useState, useMemo, useCallback } from "react";
+import { ModuleAccess } from "../../domain/entities/User";
+import { useHistory, useLocation } from "react-router-dom";
+import { defaultModuleContextState, CurrentModuleContext } from "../contexts/current-module-context";
+
+import { useAppContext } from "../contexts/app-context";
+
+export const CurrentModuleContextProvider: React.FC = ({ children }) => {
+    const history = useHistory();
+    const location = useLocation();
+    const queryParameters = useMemo(() => new URLSearchParams(location.search), [location]);
+    const moduleQueryParam = queryParameters.get("module");
+    const { currentUser } = useAppContext();
+
+    const [currentModuleAccess, setCurrentModuleAccess] = useState<ModuleAccess>(
+        defaultModuleContextState.currentModuleAccess
+    );
+
+    const [userModulesAccess, setUserModulesAccess] = useState<ModuleAccess[] | null>(null);
+
+    const changeCurrentModuleAccess = useCallback(
+        (updated: string) => {
+            const currentModuleAccess = userModulesAccess?.find(m => m.moduleName === updated);
+            if (currentModuleAccess) {
+                setCurrentModuleAccess(currentModuleAccess);
+                if (queryParameters.get("module")) {
+                    queryParameters.set("module", currentModuleAccess.moduleName);
```

Comment:
`queryParameters.set`. `queryParameters` is a memoized URLSearchParams. This should not be update inplace. It won't change the identity of the object so no part of the code will see it change (for example, the dependency of a useEffect). If that's precisely what you want, not to trigger changes, then it should be written in another way.  

A proposal: `queryParameters` should hold an object, not a `URLSearchParams`.

Reply:
Addressed in https://github.com/EyeSeeTea/glass-dev/pull/35/commits/974f1dceec2da3459feecfac63ead0a9e9916d00

Addressed this by removing queryParameter as a global variable. It is nothing but a intermediate for parsed "location.search". However, did not go ahead with making it an object, because then we would have to write the logic to parse query parameters string ourselves(could be error prone). I thought it might be a good idea to keep using  URLSearchParams to parse the query parameters as it is tried and tested. However, Please let me know if you see a downside to this, I will write remove the URLSearchParams  and write the parsing logic myself.

Reply:
No problem (Not that I wanted to remove `URLSearchParams`, it certainly does a good job on the parsing.)

If it works and the component re-when expected, leave it as-is. We can review at a later time. The general idea is that `queryParameters.set(...)` is changing an object in-place, and that's conceptually problematic in React components.